pbft/pbft.cpp

```diff
@@ -272,26 +278,32 @@ pbft::preliminary_filter_msg(const pbft_msg& msg)
 }
 
 std::shared_ptr<pbft_operation>
-pbft::setup_request_operation(const bzn_envelope& request_env, const bzn::hash_t& request_hash, const std::shared_ptr<session_base>& session)
+pbft::setup_request_operation(const bzn_envelope& request_env, const bzn::hash_t& request_hash)
 {
     const uint64_t request_seq = this->next_issued_sequence_number++;
-    auto op = this->find_operation(this->view, request_seq, request_hash);
+    auto op = this->operation_manager->find_or_construct(this->view, request_seq, request_hash, this->current_peers_ptr());
     op->record_request(request_env);
 
-    if (session)
-    {
-        op->set_session(session);
-    }
-
     return op;
 }
 
 void
 pbft::handle_request(const bzn_envelope& request_env, const std::shared_ptr<session_base>& session)
 {
+    const auto hash = this->crypto->hash(request_env);
+
+    if (session)
+    {
+        const auto hash = this->crypto->hash(request_env);
+        if (this->sessions_waiting_on_forwarded_requests.find(hash) == this->sessions_waiting_on_forwarded_requests.end())
+        {
+            this->sessions_waiting_on_forwarded_requests[hash] = session;
+        }
+    }
+
```

Comment:
Could just do this->sessions_waiting_on_forwarded_requests.insert(std::make_pair(hash, session)); 

If element exists insert will fail.